### ISSUE TITLE
Handle layout properly

### ISSUE THREAD

**Issue by GoogleCodeExporter** [melt-umn/silver#37]
```
Currently, the way Silver does layout is quite broken.

We should change it to this:

1. Introduce 'layout { ... }' as a modifier on nonterminal declarations.
2. Introduce 'layout { ... }] as a modifier for parser declarations.

The layout on a parser will specify the Initial and Final layout terminals: 
those before and after the start nonterminal is parsed.

The layout on a nonterminal will specify the Default Layout for all productions 
of that nonterminal.

Next, in the absence of a layout on a parser, all Ignore Terminals in the 
specified grammars should be used -- same behavior as today.

Finally, in the absence of a layout on a nonterminal, it should be inferred, 
somehow. Ideally the inference should mean no changes to current code. So, 
perhaps it will use all Ignore Terminals in the parser spec, as we do today.

However, perhaps a more modular inference method is necessary.  One possibility 
is to infer it based on the Ignore Terminals visible to the nonterminal 
declaration in the Silver code... but this seems fragile and might break 
things. (Consider importing a grammar that doesn't end up in some parser 
because we don't actually want to use its concrete syntax.)

Another possibility is to infer as the union of the "parents" of that 
nonterminal... that is, the other nonterminals that reach it. With the start 
nonterminal also getting a contribution from the parser's layout.

Another possibility is to introduce another declaration in the module system, 
to take care of indicating what else must be included for this grammar to be 
included in a parser.  Some sort of "concrete dependencies." Then we could use 
"all ignore terminals declared in the concrete dependencies of this grammar. 
(This nice thing about this declaration is that we would then be able to ensure 
all symbols are present in Silver, instead of once the copper compiler is run. 
In other words, this would be useful for a lot of other things than just layout 
inference.)
```

Original issue reported on code.google.com by `tux...@gmail.com` on 20 Apr 2013 at 5:25

**Comment by krame505** [comment]
I've been taking a stab at fixing this, because it has become sufficiently painful when combining multiple host languages in a single parser (as in the case of silver-ableC) due to the introduction of lexer conflicts involving ignore terminals.  We have been working around this using disambiguation functions, parser attributes and dummy terminals that flag when transitioning between meta- and object-language code, but this is awful and still breaks for some cases involving maximal munch (e.g. the C decrement operator gets parsed as a Silver comment sometimes.)

Some further thoughts here:
* Using all terminals in the specified grammar for default parser layout seems wrong: this could still lead to ambiguities between ignore terminals.  Why not just use the layout of the start nonterminal as the default here?
* I agree inferring the layout from all visible imports ("approach 1") could be fragile.  But maybe not the worst idea if the alternatives don't work out.
* Inferring the layout from all the "parents" of a nonterminal ("approach 2", I assume parent productions, not nonterminals?) means that bridge productions between host languages would still cause ignore terminals from one to "spill" into the layout of the other.  Presumably this inference approach should only consider productions exported by the nonterminal definition?  Also it seems we should also include in the default any ignore terminals exported by the nonterminal, as well.  
* Some explicit notion of "concrete imports" would certainly be nice to have.  Presumably all exports would implicitly be concrete imports, and a concrete import would also function as a regular import?  The biggest drawback to this approach is that it isn't backwards compatible, and approach 2, based on inferring from the "parents", would actually be more specific.  

So unless someone has any further thoughts on this, I think I will continue with implementing approach 2.
